README
PEXPORTS 0.46 README
============================================

Anders Norlander <anorland@hem2.passagen.se>
URL: hem2.passagen.se/anorland/

Hacked by Paul.Sokolovsky@technologist.com
URL: http://www.is.lg.ua/~paul/devel/binutils.html

Hacked by Tor Lillqvist <tml@iki.fi>

============================================

PEXPORTS is a program to extract exported symbols from a PE image
(executable). It can perform a simple check on the size of the
arguments of the exported functions, provided there is a header with
prototypes for the functions. This is useful when you want the
decorated function names for functions using the stdcall calling
convention. GCC is used to do the preprocessing so it must be in your
path.

Note that the Windows version uses ';' as path separator,
while if built for Cygwin (or *nix) it uses ':'.

Command line options:
=====================
        -h <header> parse header
        -o print function ordinals
        -p <preprocessor> set preprocessor
        -v verbose mode

Header files are searched for in the following directories:
1. Current directory
2. Directories in C_INCLUDE_PATH
3. Directories in CPLUS_INCLUDE_PATH

NOTE: The header parser is *very* primitive, and might be of
questionable usefulness. It only tries to find function prototypes and
check the number of arguments a function expects. It is NOT a complete
C parser, there are probably many conditions when it will fail
(especially complex parameter types), although I it works fine for me.
Please do not report bugs, but feel free to send patches.

RELEASE 0.46
=================
* Corrects two potential segmentation fault bugs.
* Now supports building "out-of-source", to facilitate parallel builds
  for differing hosts; (verified for mingw32, 32-bit and 64-bit linux).
* Now uses the GNU (autoconf managed) build procedure; run configure
  before make, when building, ("in-source" or "out-of-source").

RELEASE 0.45
=================
* Incorporate patches from 0.44-1-mingw32 release
* Reinstate binary distribution for mingw32 host only;
  (distributed source is host-agnostic).

RELEASE 0.44
=================
* Handle also 64-bit executables. Make it work also if built as 64-bit
  code. Remove all gcc -Wall warnings. Make it compilable also with
  MSVC. Disable Wow64 file system redirection when running as a 32-bit
  process on 64-bit Windows. Distribute just sources.

RELEASE 0.43
=================
* There were bug which led to wrong subcategorizing of symbols as
  code/data. I thought it was fixed in version on the site, but it turns
  out that almost year there was wrong version. I greatly apologize to
  everyone whom it cause problems and confusion.

RELEASE 0.42
=================
* Data/non-data symbols are now distinguished.

RELEASE 0.41
=================
* The header parser now accepts all kinds of parameters.

RELEASE 0.4
=================
* Function pointer parameters are now handled
* Handling of function attributes improved
* It is no longer always necessary to include windows.h for headers
  that required it but did not include it themselves.

RELEASE 0.3
=================
* Completely rewritten parser (the previous one was *very* bad).
  It is now possible to generate .DEF files for windows system
  dlls (kernel32,user32,gdi32,shell32 etc)
* Enhanced symbol handling (symbols are sorted in a tree).

RELEASE 0.2
=================
* Fixed bug with unnamed parameters that are pointers.
* Extra whitespace is no longer printed
* Binary versions available for mingw32 and cygwin32

RELEASE 0.1
=================
* Initial release.
Pexports, Copyright (C) 1998 Anders Norlander
This program has ABSOLUTELY NO WARRANTY; This is free software, and you are
welcome to redistribute it under certain conditions; see COPYING
for details.


ChangeLog
2015-09-08  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Prepare, tag, and publish 0.47 release.

	* configure.ac: Update copyright notice.
	(AC_INIT): Increment PACKAGE_VERSION assignment to 0.47

	* pexports.c: Update copyright notice.

	* README: Add release notes.
	* All files: Tagged "v0_47" for release.

2015-09-07  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Avoid uint32_t difference overflow in 64-bit pointer adjustment.

	* pexports.c (rva_to_ptr): Do not compute uint32_t offset differences;
	apply each offset adjustment individually to the base pointer.

	* pexports.h: Add '#pragma ms_struct on', to keep structures aligned
	per MSVC convention.

2015-09-07  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Eliminate Microsoft inspired obfuscated typedef insanity.

	* pexports.c pexports.h (BYTE, WORD, DWORD): Delete typedefs;
	respectively replace all corresponding usage instances with...
	(uint8_t, uint16_t, uint32_t): ...these ISO-C standard types.
	(PVOID, PIMAGE_DOS_HEADER, PIMAGE_FILE_HEADER, PIMAGE_SECTION_HEADER)
	(PIMAGE_DATA_DIRECTORY, PIMAGE_EXPORT_DIRECTORY, PIMAGE_NT_HEADERS32)
	(PIMAGE_NT_HEADERS64, PIMAGE_OPTIONAL_HEADER32)
	(PIMAGE_OPTIONAL_HEADER64): Delete these non-standard style pointer
	reference typedefs; replace with conventional pointer references to
	the underlying data type; where these are structural types, delete
	initial underscores from the corresponding structure tag names.
	(HMODULE): Likewise; replace with pointer to void.

2013-05-30  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Prepare, tag, and publish 0.46 release.

	* README: Add release notes.
	* All files: Tagged "v0_46" for release.

2013-05-30  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Assemble licence and documentation tarballs for distribution.

	* Makefile.in (dist): Add dependencies on...
	(licdist, docdist): ...these new build goals; specify them based on...
	(LIC_DISTFILES, DOC_DISTFILES): ...these new macros; define them as
	the respective list of files to include in each tarball.

2013-05-29  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Force a remake, on package version updates.

	* Makefile.in (pexports.OBJEXT): Make it depend on Makefile; this
	indirectly captures changes, when configure.ac is updated.

2013-05-29  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Update program identification, via configuration process.

	* pexports.c (main): Update identification text; remove references...
	(VER_MAJOR, VER_MINOR): ...to these; combine them by substituting...
	(PACKAGE_VERSION_STRING): ...this new macro; also amend instructions
	for reporting bugs, as directed at URL identified by...
	(PACKAGE_BUG_REPORT): ...this additional new macro.

	* pexports.h (VER_MAJOR, VER_MINOR): Obsolete macros; delete them.

	* Makefile.in (PACKAGE_DEFS): New macro; defined in terms of these...
	(PACKAGE_VERSION_STRING, PACKAGE_BUG_REPORT_URL, PACKAGE_BUG_REPORT):
	...these additional new macros, defined by AC_SUBST from configure.ac,
	and subsequently propagated to pexports.c, as an addendum to...
	(CDEPFLAGS): ...this.

2013-05-29  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Autoconfiscate.

	* configure.ac Makefile.in: New files.
	* Makefile: Now auto-generated at configure time; remove it.
	* .hgignore: Add configure, Makefile, autom4te.cache and build dir.

2013-05-29  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Integrate build-aux utilities as a submodule.

	* .gitmodules .hgsub .hgsubstate: New files.

2013-05-28  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Designate files which mercurial should not track.

	* .hgignore: New file.

2013-05-28  Keith Marshall  <keithmarshall@users.sourceforge.net>

	Correct some obfuscated typedef portability issues.

	* pexports.c: Include inttypes.h; it permits use of...
	(PRIdPTR): ...this ISO-C99 standard macro, in preference to...
	(INT_PTR_FORMAT): ...this non-standard contrivance; similarly...
	(intptr_t): ...substitute this ISO-C99 standard typedef...
	(INT_PTR): ...for this non-standard type.

	* pexports.h: Include stdint.h; typedef...
	(BYTE): ...this, as explicitly equivalent to uint8_t.
	(WORD): ...this, as explicitly equivalent to uint16_t.
	(DWORD): ...this, as explicitly equivalent to uint32_t.
	(INT_PTR, INT_PTR_FORMAT): No longer required; delete them.
	(ULONGLONG): Explicitly typedef as equivalent to uint64_t; although
	'unsigned long long' seems a better fit, our present usage requires
	the explicit 64-bit association, and use of this standard type allows
	us to eliminate all references to Microsoft's non-standard __int64.
	(LONG): Explicitly typedef as equivalent to int32_t; technically, this
	is incorrect, but misuse elsewhere requires the explicit 32-bit type,
	which is not guaranteed by ISO-C99's standard 'long' type.

2011-09-14  Daniel Collins  <solemnwarning@solemnwarning.net>

	Avoid segmentation faults in forwarder function checks.

	* pexports.h (dump_exports): Adjust prototype to accommodate new
	'exports_size' argument; this is to be assigned and passed from...
	* pexports.c (main): ...here; assigned value is determined from the
	export table size recorded in the PE file image, whence passed to...
	(dump_exports): ...here; it is used to avoid crashing or corruption
	when the export table isn't at the end of its section.  Also check
	for, and filter out duplicated symbol reports.

2009-09-15  Charles Wilson  <cwilso11@users.sourceforge.net>

	Slight build improvements.

	* README: Update.
	* Makefile: Rewrite dist rule to create a .tar.lzma instead
	of a .zip (and don't create a tarbomb). 

2009-09-15  Charles Wilson  <cwilso11@users.sourceforge.net>

	Package 0.44-1-mingw32 fixes.

	* pexports.c (main): Fix typo in usage statement.
	* pexports.h: Ensure ULONGLONG is defined.
	* Makefile: Force use of flex and bison, not lex and yacc.

2008-08-31  Tor Lillqvist  <tml@novell.com>


pexports.h
/*
 pexports - a program to extract exported symbols from a Portable
 Executable (PE) file.

 Copyright (C) 1998 Anders Norlander

 pexports is distributed under the GNU General Public License and
 has absolutely NO WARRANTY.

 pexports will work only on intel machines.
*/
#ifndef _pexports_h
#define _pexports_h

#pragma ms_struct on

#include <stdio.h>
#include <stdint.h>

#include "str_tree.h"

#define FALSE (int32_t)(0)
#define TRUE  (int32_t)(1)

/* PE structures */
typedef struct IMAGE_DATA_DIRECTORY {
  uint32_t   VirtualAddress;
  uint32_t   Size;
} IMAGE_DATA_DIRECTORY;

typedef struct IMAGE_FILE_HEADER {
  uint16_t   Machine;
  uint16_t   NumberOfSections;
  uint32_t   TimeDateStamp;
  uint32_t   PointerToSymbolTable;
  uint32_t   NumberOfSymbols;
  uint16_t   SizeOfOptionalHeader;
  uint16_t   Characteristics;
} IMAGE_FILE_HEADER;

#define IMAGE_FILE_MACHINE_I386  0x014c
#define IMAGE_FILE_MACHINE_IA64  0x0200
#define IMAGE_FILE_MACHINE_AMD64 0x8664

typedef struct IMAGE_OPTIONAL_HEADER32 {
  uint16_t   Magic;
  uint8_t    MajorLinkerVersion;
  uint8_t    MinorLinkerVersion;
  uint32_t   SizeOfCode;
  uint32_t   SizeOfInitializedData;
  uint32_t   SizeOfUninitializedData;
  uint32_t   AddressOfEntryPoint;
  uint32_t   BaseOfCode;
  uint32_t   BaseOfData;
  uint32_t   ImageBase;
  uint32_t   SectionAlignment;
  uint32_t   FileAlignment;
  uint16_t   MajorOperatingSystemVersion;
  uint16_t   MinorOperatingSystemVersion;
  uint16_t   MajorImageVersion;
  uint16_t   MinorImageVersion;
  uint16_t   MajorSubsystemVersion;
  uint16_t   MinorSubsystemVersion;
  uint32_t   Reserved1;
  uint32_t   SizeOfImage;
  uint32_t   SizeOfHeaders;
  uint32_t   CheckSum;
  uint16_t   Subsystem;
  uint16_t   DllCharacteristics;
  uint32_t   SizeOfStackReserve;
  uint32_t   SizeOfStackCommit;
  uint32_t   SizeOfHeapReserve;
  uint32_t   SizeOfHeapCommit;
  uint32_t   LoaderFlags;
  uint32_t   NumberOfRvaAndSizes;
  IMAGE_DATA_DIRECTORY DataDirectory[16];
} IMAGE_OPTIONAL_HEADER32;

typedef struct IMAGE_OPTIONAL_HEADER64 {
  uint16_t   Magic;
  uint8_t    MajorLinkerVersion;
  uint8_t    MinorLinkerVersion;
  uint32_t   SizeOfCode;
  uint32_t   SizeOfInitializedData;
  uint32_t   SizeOfUninitializedData;
  uint32_t   AddressOfEntryPoint;
  uint32_t   BaseOfCode;
  uint64_t   ImageBase;
  uint32_t   SectionAlignment;
  uint32_t   FileAlignment;
  uint16_t   MajorOperatingSystemVersion;
  uint16_t   MinorOperatingSystemVersion;
  uint16_t   MajorImageVersion;
  uint16_t   MinorImageVersion;
  uint16_t   MajorSubsystemVersion;
  uint16_t   MinorSubsystemVersion;
  uint32_t   Win32VersionValue;
  uint32_t   SizeOfImage;
  uint32_t   SizeOfHeaders;
  uint32_t   CheckSum;
  uint16_t   Subsystem;
  uint16_t   DllCharacteristics;
  uint64_t   SizeOfStackReserve;
  uint64_t   SizeOfStackCommit;
  uint64_t   SizeOfHeapReserve;
  uint64_t   SizeOfHeapCommit;
  uint32_t   LoaderFlags;
  uint32_t   NumberOfRvaAndSizes;
  IMAGE_DATA_DIRECTORY DataDirectory[16];
} IMAGE_OPTIONAL_HEADER64;

typedef struct IMAGE_NT_HEADERS32 {
  char Signature[4];
  IMAGE_FILE_HEADER FileHeader;
  IMAGE_OPTIONAL_HEADER32 OptionalHeader;
} IMAGE_NT_HEADERS32;

typedef struct IMAGE_NT_HEADERS64 {
  char Signature[4];
  IMAGE_FILE_HEADER FileHeader;
  IMAGE_OPTIONAL_HEADER64 OptionalHeader;
} IMAGE_NT_HEADERS64;

typedef struct IMAGE_SECTION_HEADER {
  uint8_t    Name[8];
  union {
    uint32_t PhysicalAddress;
    uint32_t VirtualSize;
  } Misc;
  uint32_t   VirtualAddress;
  uint32_t   SizeOfRawData;
  uint32_t   PointerToRawData;
  uint32_t   PointerToRelocations;
  uint32_t   PointerToLinenumbers;
  uint16_t   NumberOfRelocations;
  uint16_t   NumberOfLinenumbers;
  uint32_t   Characteristics;
} IMAGE_SECTION_HEADER;

#define IMAGE_SCN_CNT_CODE 0x00000020

typedef struct IMAGE_EXPORT_DIRECTORY {
  uint32_t   Characteristics;
  uint32_t   TimeDateStamp;
  uint16_t   MajorVersion;
  uint16_t   MinorVersion;
  uint32_t   Name;
  uint32_t   Base;
  uint32_t   NumberOfFunctions;
  uint32_t   NumberOfNames;
  uint32_t   AddressOfFunctions;
  uint32_t   AddressOfNames;
  uint32_t   AddressOfNameOrdinals;
} IMAGE_EXPORT_DIRECTORY;

typedef struct IMAGE_DOS_HEADER {
  uint16_t   e_magic;
  uint16_t   e_cblp;
  uint16_t   e_cp;
  uint16_t   e_crlc;
  uint16_t   e_cparhdr;
  uint16_t   e_minalloc;
  uint16_t   e_maxalloc;
  uint16_t   e_ss;
  uint16_t   e_sp;
  uint16_t   e_csum;
  uint16_t   e_ip;
  uint16_t   e_cs;
  uint16_t   e_lfarlc;
  uint16_t   e_ovno;
  uint16_t   e_res[4];
  uint16_t   e_oemid;
  uint16_t   e_oeminfo;
  uint16_t   e_res2[10];
  int32_t    e_lfanew;
} IMAGE_DOS_HEADER;

IMAGE_SECTION_HEADER *find_section( uint32_t );
IMAGE_DOS_HEADER *load_pe_image( const char * );

void *rva_to_ptr( uint32_t );
void dump_exports( uint32_t, uint32_t );

#define ADD_FUNCTION(nm,n) str_tree_add(&symbols, nm, (void*)(intptr_t)(n))
extern str_tree *symbols;

#endif /* _pexports_h */


pexports.c
#include <string.h>
#include <stdlib.h>
#include <inttypes.h>
#include <sys/stat.h>
#include <errno.h>

#include "str_tree.h"
#include "pexports.h"

#ifndef PATH_MAX
#define PATH_MAX 260
#endif

#ifdef _WIN32
#define PATH_SEPARATOR ';'
#else
#define PATH_SEPARATOR ':'
#endif

/* get pointer to section header n */
#define IMAGE_SECTION_HDR(n) \
  ((IMAGE_SECTION_HEADER *)((char *) nt_hdr32 \
    + 4 + sizeof(IMAGE_FILE_HEADER) + nt_hdr32->FileHeader.SizeOfOptionalHeader \
    + n * sizeof(IMAGE_SECTION_HEADER)) \
  )

/* convert relative virtual address to a useable pointer */
#define RVA_TO_PTR(rva,type) ((type)(rva_to_ptr((uint32_t)(rva))))

typedef struct str_list {
  char *s;
  struct str_list *next;
} str_list;

extern void yyparse(void);

static void *xmalloc(size_t count);
static void add_path_list(char *path);
static const char *find_file(const char *name);
static void str_list_add(str_list **head, const char *s);
static void parse_headers();
static void dump_symbol(char *name, int ord, uint32_t rva);

static const char mz_sign[2] = {'M','Z'};
static const char pe_sign[4] = {'P','E','\0','\0'};
static const char exp_sign[6] = {'.','e','d','a','t','a'};

static IMAGE_DOS_HEADER   *dos_hdr;
static IMAGE_NT_HEADERS32 *nt_hdr32;
static IMAGE_NT_HEADERS64 *nt_hdr64;

static char *filename = NULL;
static char *program_name;
static char *cpp = "gcc -E -xc-header";

str_tree *symbols = NULL;
static str_list *inc_path = NULL;
static str_list *header_files = NULL;

static int verbose = 0;
static int ordinal_flag = 0;

extern FILE *yyin;

int
main(int argc, char *argv[])
{
  IMAGE_SECTION_HEADER *section;
  uint32_t exp_rva, exp_size;
  int i;

# if defined(_WIN32) && !defined(_WIN64)
  /*
   * If running on 64-bit Windows and built as a 32-bit process, try
   * disable Wow64 file system redirection, so that we can open DLLs
   * in the real system32 folder if requested.
   */
  void *old_redirection;
  void *kernel32;

  extern __declspec(dllimport) void __stdcall *GetModuleHandleA(char *name);
  extern __declspec(dllimport) void __stdcall *GetProcAddress(void *module, char *name);

  int32_t (__stdcall *pWow64DisableWow64FsRedirection) (void **old_value);

  kernel32 = GetModuleHandleA("kernel32.dll");
  pWow64DisableWow64FsRedirection = GetProcAddress(kernel32, "Wow64DisableWow64FsRedirection");

  if (pWow64DisableWow64FsRedirection)
    pWow64DisableWow64FsRedirection(&old_redirection);
# endif

  program_name = argv[0];

  /* add standard include paths */
  add_path_list(getenv("C_INCLUDE_PATH"));
  add_path_list(getenv("CPLUS_INCLUDE_PATH"));
#if 0
  /* since when does PATH have anything to do with headers? */
  add_path_list(getenv("PATH"));
#endif

  /* parse command line */
  for ( i = 1; i < argc; i++ )
    {
      if (argv[i][0] == '-')
        {
          switch (argv[i][1])
            {
            case 'v':
              verbose = 1;
              break;
            case 'o':
              ordinal_flag = 1;
              break;
            case 'h':
              if (!argv[++i])
                {
                  fprintf(stderr, "-h requires an argument\n");
                  return 1;
                }
              str_list_add(&header_files, argv[i]);
              break;
            case 'p':
              if (!argv[++i])
                {
                  fprintf(stderr, "-p requires an argument\n");
                  return 1;
                }
              cpp = argv[i];
              break;
            default:
              fprintf(stderr, "%s: Unknown option: %s\n",
                      program_name, argv[i]);
              return 1;
            }
        }
      else
        filename = argv[i];
    }

  if (filename == NULL)
    {
      printf(
	"PExports %s; Originally written 1998, Anders Norlander\n"
   	"Updated 1999, Paul Sokolovsky, 2008, Tor Lillqvist, 2013, 2015, Keith Marshall\n"
	"Copyright (C) 1998, 1999, 2008, 2013, 2015, MinGW.org Project\n\n"
	"This program is free software; you may redistribute it under the terms of\n"
	"the GNU General Public License.  This program has absolutely no warranty.\n"

	"\nUsage: %s [-v] [-o] [-h header] [-p preprocessor] dll\n"
	"  -h\tparse header\n"
	"  -o\tprint ordinals\n"
	"  -p\tset preprocessor program\n"
	"  -v\tverbose mode\n"
	"\nReport bugs as directed at %s\n",
	PACKAGE_VERSION_STRING, program_name, PACKAGE_BUG_REPORT);
      return 1;
    }

  /* parse headers and build symbol tree */
  parse_headers();

  /* load file */
  if( (dos_hdr = load_pe_image(filename)) == NULL )
    {
      fprintf(stderr, "%s: %s: could not load PE image\n",
              program_name, filename);
      return 1;
    }

  nt_hdr32 = (IMAGE_NT_HEADERS32 *) ((char *) dos_hdr + dos_hdr->e_lfanew);
  nt_hdr64 = (IMAGE_NT_HEADERS64 *) nt_hdr32;

  if (nt_hdr32->FileHeader.Machine == IMAGE_FILE_MACHINE_I386) {
    exp_rva = nt_hdr32->OptionalHeader.DataDirectory[0].VirtualAddress;
    exp_size = nt_hdr32->OptionalHeader.DataDirectory[0].Size;
  }else{
    exp_rva = nt_hdr64->OptionalHeader.DataDirectory[0].VirtualAddress;
    exp_size = nt_hdr64->OptionalHeader.DataDirectory[0].Size;
  }

  if (verbose)
    {
      for (i = 0; i < nt_hdr32->FileHeader.NumberOfSections; i++)
        {
	  section = IMAGE_SECTION_HDR(i);
          printf("; %-8.8s: RVA: %08x, File offset: %08x\n",
                 section->Name,
                 section->VirtualAddress,
                 section->PointerToRawData);
        }
    }

  /* Look for export section */
  for (i = 0; i < nt_hdr32->FileHeader.NumberOfSections; i++)
    {
      section = IMAGE_SECTION_HDR(i);
      if (memcmp(section->Name, exp_sign, sizeof(exp_sign)) == 0)
        dump_exports(section->VirtualAddress, exp_size);
      else if ((exp_rva >= section->VirtualAddress) &&
          (exp_rva < (section->VirtualAddress + section->SizeOfRawData)))
        dump_exports(exp_rva, exp_size);
    }

  free(dos_hdr);
  return 0;
}

/* dump exported symbols on stdout */
void
dump_exports(uint32_t exports_rva, uint32_t exports_size)
{
  IMAGE_SECTION_HEADER *section;
  IMAGE_EXPORT_DIRECTORY *exports;
  char *export_name;
  uint16_t *ordinal_table;
  uint32_t *name_table;
  uint32_t *function_table;
  int i;
  static int first = 1;

  section = find_section(exports_rva);

  if (verbose)
    printf("; Reading exports from section: %s\n",
            section->Name);

  exports = RVA_TO_PTR(exports_rva, IMAGE_EXPORT_DIRECTORY *);

  /* set up various pointers */
  export_name = RVA_TO_PTR(exports->Name, char *);
  name_table = RVA_TO_PTR(exports->AddressOfNames, uint32_t *);
  ordinal_table = RVA_TO_PTR(exports->AddressOfNameOrdinals, uint16_t *);
  function_table = RVA_TO_PTR(exports->AddressOfFunctions, void *);

  if (verbose)
    {
      printf("; Export table: %s\n", export_name);
      printf("; Ordinal base: %d\n", exports->Base);
      printf("; Ordinal table RVA: %08x\n",
              exports->AddressOfNameOrdinals);
      printf("; Name table RVA: %07x\n",
	     exports->AddressOfNames);
      printf("; Export address table RVA: %08x\n",
             exports->AddressOfFunctions);
    }

  if (first)
    {
      printf("LIBRARY %s\n", export_name);
      printf("EXPORTS\n");
      first = 0;
    }
  else
      printf("; LIBRARY %s\n", export_name);

  for (i = 0; i < exports->NumberOfNames; i++)
    {
      dump_symbol(RVA_TO_PTR(name_table[i],char*),
                  ordinal_table[i] + exports->Base,
                  function_table[ordinal_table[i]]);

      int f_off = ordinal_table[i];

      if(function_table[f_off] >= exports_rva && function_table[f_off] < (exports_rva + exports_size) && verbose) {
        printf(" ; Forwarder (%s)", RVA_TO_PTR(function_table[f_off], char*));
      }

      printf("\n");
    }

  for (i = 0; i < exports->NumberOfFunctions; i++)
    {
      if ( (function_table[i] >= exports_rva) &&
           (function_table[i] < (exports_rva + exports_size)))
        {
          int name_present = 0, n;

          for(n = 0; n < exports->NumberOfNames; n++) {
            if(ordinal_table[n] == i) {
              name_present = 1;
              break;
            }
          }

          if(!name_present) {
            dump_symbol(strchr(RVA_TO_PTR(function_table[i],char*), '.')+1, i + exports->Base, function_table[i]);

            printf(" ; WARNING: Symbol name guessed from forwarder (%s)\n", RVA_TO_PTR(function_table[i], char*));
          }
        }
    }
}

static void
dump_symbol(char *name, int ord, uint32_t rva)
{
  char s[256];
  str_tree *symbol = str_tree_find(symbols, name);
  /* if a symbol was found, emit size of stack */
  if (symbol)
    sprintf(s, "%s@%"PRIdPTR, name, (intptr_t)(symbol->extra));
  else
    sprintf(s, "%s", name);

  /* output ordinal */
  if (ordinal_flag)
    printf("%-24s\t@%d", s, ord);
  else
    printf("%s", s);

  {
    IMAGE_SECTION_HEADER *s = find_section(rva);

    /* Stupid msvc doesn't have .bss section, it spews uninitilized data
     * to no section
     */
    if (!s) { printf(" DATA"); if (verbose) printf (" ; no section"); }
    else
    {
      if (!(s->Characteristics&IMAGE_SCN_CNT_CODE)) printf(" DATA");
      if (verbose) printf (" ; %.8s",s->Name);
    }
  }

  if (verbose)
    printf(" ; RVA %08x", rva);
}

/* get section to which rva points */
IMAGE_SECTION_HEADER *find_section(uint32_t rva)
{
  int i;
  IMAGE_SECTION_HEADER *section = IMAGE_SECTION_HDR(0);
  for (i = 0; i < nt_hdr32->FileHeader.NumberOfSections; i++, section++)
    if ((rva >= section->VirtualAddress) &&
        (rva <= (section->VirtualAddress + section->SizeOfRawData)))
      return section;
  return NULL;
}

/* convert rva to pointer into loaded file */
void *
rva_to_ptr(uint32_t rva)
{
  IMAGE_SECTION_HEADER *section = find_section(rva);
  return (section->PointerToRawData != 0)
    ? (char *)(dos_hdr) + rva - section->VirtualAddress + section->PointerToRawData
    : NULL;
}

/* Load a portable executable into memory */
IMAGE_DOS_HEADER *load_pe_image(const char *filename)
{
#ifdef _MSC_VER
  struct _stat32 st;
#else
  struct stat st;
#endif
  IMAGE_DOS_HEADER *phdr;
  FILE *f;

  if (stat(filename, &st) == -1)
    {
      perror("stat");
      return NULL;
    }

  phdr = (IMAGE_DOS_HEADER *) xmalloc(st.st_size);

  f = fopen(filename, "rb");

  if (f == NULL)
    {
      perror("fopen");
      free(phdr);
      return NULL;
    }

  if (fread(phdr, st.st_size, 1, f) != 1)
    {
      perror("fread");
      free(phdr);
      phdr = NULL;
    }
  else if (memcmp(mz_sign, phdr, sizeof(mz_sign)) != 0)
    {
      fprintf(stderr, "No MZ signature\n");
      free(phdr);
      phdr = NULL;
    }
  else if (memcmp((char *) phdr + phdr->e_lfanew, pe_sign, sizeof(pe_sign)) != 0)
    {
      fprintf(stderr, "No PE signature\n");
      free(phdr);
      phdr = NULL;
    }

  fclose(f);
  return phdr;
}

/* parse headers to build symbol tree */
void
parse_headers()
{
#ifdef _MSC_VER
#define popen _popen
#define pclose _pclose
#endif

  str_list *header;
  char *cpp_cmd;
  int len;
  FILE *f;

  header = header_files;
  if (!header)
    return;

  /* construct command line */
  cpp_cmd = strdup(cpp);
  if (cpp_cmd == NULL)
    {
      fprintf(stderr, "%s: out of memory\n", program_name);
      exit(1);
    }
  len = strlen(cpp_cmd);

  while (header)
    {
      const char *fullname = find_file(header->s);
      int tmp;
      if (fullname == NULL)
        {
          perror(header->s);
          exit(1);
        }
      tmp = strlen(fullname) + 10;
      cpp_cmd = realloc(cpp_cmd, len + tmp);
      if (cpp_cmd == NULL)
        {
          fprintf(stderr, "%s: out of memory\n", program_name);
          exit(1);
        }
      if (!header->next)
        sprintf(cpp_cmd + len, " %s", fullname);
      else
        sprintf(cpp_cmd + len, " -include %s", fullname);
      len += tmp;
      header = header->next;
    }
  cpp_cmd[len] = '\0';

  if (verbose)
    printf("; %s\n", cpp_cmd);

  /* Run preprocessor.
     Note: CRTDLL messes up stdout when popen is called so
     if you try to pipe output through another program
     with | it will hang. Redirect it to a file instead
     and pass that file to the program (more,less or whatever).
     This does not apply to cygwin.
  */
  f = popen(cpp_cmd, "r");
  free(cpp_cmd);
  if (f == NULL)
    {
      fprintf(stderr, "%s: %s: could not execute\n", program_name, cpp_cmd);
      exit(1);
    }
  yyin = f;
  yyparse();
  pclose(f);
}

/* allocate memory; abort on failure */
static void *xmalloc(size_t count)
{
  void *p = malloc(count);
  if (p == NULL)
    {
      fprintf(stderr, "%s: out of memory\n", program_name);
      exit(1);
    }
  return p;
}

/* add string to end of list */
static void
str_list_add(str_list **head, const char *s)
{
  str_list *node = xmalloc(sizeof(str_list));
  node->s = strdup(s);
  node->next = NULL;
  if (!*head)
    {
      *head = node;
    }
  else
    {
      str_list *p = *head;
      while (p->next)
        p = p->next;
      p->next = node;
    }
}

/* find a file in include path */
static const char *
find_file(const char *name)
{
  static char fullname[PATH_MAX];
  FILE *f = fopen(name, "r");
  str_list *path = inc_path;
  if (f != NULL)
    {
      fclose(f);
      return name;
    }
  while (path)
    {
      strcpy(fullname, path->s);
      strcat(fullname, "/");
      strcat(fullname, name);
      f = fopen(fullname, "r");
      if (f != NULL)
        {
          fclose(f);
          return fullname;
        }
      path = path->next;
    }
  errno = ENOENT;
  return NULL;
}

/* add a environment-style path list to list of include paths */
static void
add_path_list(char *path)
{
  char *p = path;
  if (!p)
    return;
  while (*p)
    {
      if (*p == PATH_SEPARATOR)
        {
          *p = '\0';
          str_list_add(&inc_path, path);
          path = p + 1;
          *p = PATH_SEPARATOR;
        }
      p++;
    }
  if (p[-1] != PATH_SEPARATOR)
    str_list_add(&inc_path, path);
}
